libsplinter/src/store/sqlite.rs

```diff
@@ -0,0 +1,54 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use diesel::{
+    r2d2::{ConnectionManager, Pool},
+    sqlite::SqliteConnection,
+};
+
+use super::StoreFactory;
+
+/// A `StoryFactory` backed by a SQLite database.
+pub struct SqliteStoreFactory {
+    pool: Pool<ConnectionManager<SqliteConnection>>,
+}
+
+impl SqliteStoreFactory {
+    pub fn new(pool: Pool<ConnectionManager<SqliteConnection>>) -> Self {
```

Comment:
If we are concerned about this diesel/r2d2 leakage, we could provide ::new() and ::new_with_pool() and then protect ::new_path_pool() with a feature flag "expose-diesel-api" (not that exact feature name though).

Reply:
Maybe sqlite-expose-diesel-pool, sqlite-expose-external-apis, or sqlite-pool-integration

Reply:
If we took the approach of hiding pool-integration behind a more specific feature, we could probably also get away with DieselStoreFactore<C> where C: Connection and only expose DieselStoreFactory if the feature is enabled.

Reply:
I don't think we can get away with a generic `DieselStoreFactore<C: Connection>`, because the stores don't have generic implementations for `C: Connection`; they have implementations for  `PgConnection` and `SqliteConnection`. The compiler needs to know which type of connection is to pick the right trait implementations for the stores.

As far as the diesel/r2d2 API leakage, I'm not personally concerned about it as long as we re-export the diesel/r2d2 elements that are exposed (to effectively pin them to the same version splinter uses).

Reply:
Do these constructors need to be crate public?

Reply:
They do if we want to allow a single connection pool to be used for stores that aren't internal to Splinter, which is a requirement per @vaporos.